README.md
# axios-retry

Axios plugin that intercepts failed requests and retries them whenever possible.

## Installation

```bash
npm install @vtex/axios-concurrent-retry
```

## Usage

```js
// CommonJS
// const axiosRetry = require('@vtex/axios-concurrent-retry');

// ES6
import axiosRetry from '@vtex/axios-concurrent-retry';

axiosRetry(axios, { retries: 3 });

axios.get('http://example.com/test') // The first request fails and the second returns 'ok'
  .then(result => {
    result.data; // 'ok'
  });

// Exponential back-off retry delay between requests
axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay});

// Custom retry delay
axiosRetry(axios, { retryDelay: (retryCount) => {
  return retryCount * 1000;
}});

// Works with custom axios instances
const client = axios.create({ baseURL: 'http://example.com' });
axiosRetry(client, { retries: 3 });

client.get('/test') // The first request fails and the second returns 'ok'
  .then(result => {
    result.data; // 'ok'
  });

// Allows request-specific configuration
client
  .get('/test', {
    'axios-retry': {
      retries: 0
    }
  })
  .catch(error => { // The first request fails
    error !== undefined
  });
```

**Note:** the plugin interprets the request timeout as a global value, so it is not used for each retry but for the whole request lifecycle.

## Options

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| retries | `Number` | 3 | The number of times to retry before failing |
| retryCondition | `Function` | `isNetworkOrIdempotentRequestError` | A callback to further control if a request should be retried.  By default, it retries if it is a network error or a 5xx error on an idempotent request (GET, HEAD, OPTIONS, PUT or DELETE). |
| retryDelay | `Function` | `0` | A callback to further control the delay between retried requests. By default there is no delay between retries. Another option is exponentialDelay ([Exponential Backoff](https://developers.google.com/analytics/devguides/reporting/core/v3/errors#backoff)). The function is passed `retryCount` and `error`. |

## Testing

Clone the repository and execute:

```bash
npm test
```

## Contribute

1. Fork it: `git clone https://github.com/softonic/axios-retry.git`
2. Create your feature branch: `git checkout -b feature/my-new-feature`
3. Commit your changes: `git commit -am 'Added some feature'`
4. Check the build: `npm run build`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request :D


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

### Fixed
- Added `exponentialDelay` to `IAxiosRetry` interface

## [3.0.2] - 2018-02-09

### Added
- Now isRetryableError method is accessible.
- Added delayStrategy option to be able to have exponential backoff for successive retries.

## [3.0.1] - 2017-08-16

### Fixed
- Fixed first request time not being taken into account in timeout across retries.
- Fixed negative timeouts being passed to XHR (browsers), causing that no timeout was applied.
- Fixed safe methods and idempotent errors not being retried on unknown network errors.

## [3.0.0] - 2017-08-13

### Changed
- Retried errors on idempotent requests (5xx with get, head, options, put and delete) by default,
along with safe network errors.
- Moved some hard-coded conditions to the default `retryCondition` function so users can define a
custom function that overwrites them. The conditions that verify that the error is not a timeout or
an unsafe network error have been moved to `isNetworkError`.

### Added
- Added additional pre-defined retry conditions: `isSafeRequestError`, `isIdempotentRequestError`.

## [2.0.1] - 2017-06-19

### Fixed
- Removed dependency from the `package.json` file.

## [2.0.0] - 2017-06-15

### Changed
- Now the configured timeout in Axios is not for each retry request but for the whole request lifecycle.

## [1.3.1] - 2017-06-19

### Fixed
- Removed dependency from the `package.json` file.

## [1.3.0] - 2017-06-15

### Added
- Allowed per-request configuration using the `axios-retry` namespace.


spec/index.spec.js
import http from 'http'
import nock from 'nock'
import axios from 'axios'
import {
  default as axiosRetry,
  isNetworkError,
  isSafeRequestError,
  isIdempotentRequestError,
  exponentialDelay,
  isRetryableError,
} from '../es/index'

const NETWORK_ERROR = new Error('Some connection error')
NETWORK_ERROR.code = 'ECONNRESET'

function setupResponses(client, responses) {
  const configureResponse = () => {
    const response = responses.shift()
    if (response) {
      response()
    }
  }
  client.interceptors.response.use(result => {
    configureResponse()
    return result
  }, error => {
    configureResponse()
    return Promise.reject(error)
  })
  configureResponse()
}

describe('axiosRetry(axios, { retries, retryCondition })', () => {
  afterEach(() => {
    nock.cleanAll()
    nock.enableNetConnect()
  })

  describe('when the response is successful', () => {
    it('should resolve with it', done => {
      const client = axios.create()
      setupResponses(client, [
        () => nock('http://example.com').get('/test').reply(200, 'It worked!'),
      ])

      axiosRetry(client, { retries: 0 })

      client.get('http://example.com/test').then(result => {
        expect(result.status).toBe(200)
        done()
      }, done.fail)
    })
  })

  describe('when the response is an error', () => {
    it('should check if it satisfies the `retryCondition`', (done) => {
      const client = axios.create()
      setupResponses(client, [
        () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
        () => nock('http://example.com').get('/test').reply(200, 'It worked!'),
      ])

      const retryCondition = (error) => {
        expect(error).toBe(NETWORK_ERROR)
        done()
        return false
      }

      axiosRetry(client, { retries: 1, retryCondition })

      client.get('http://example.com/test').catch(() => {})
    })

    describe('when it satisfies the retry condition', () => {
      it('should resolve with a successful retry', (done) => {
        const client = axios.create()
        setupResponses(client, [
          () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
          () => nock('http://example.com').get('/test').reply(200, 'It worked!'),
        ])

        axiosRetry(client, { retries: 1, retryCondition: () => true })

        client.get('http://example.com/test').then(result => {
          expect(result.status).toBe(200)
          done()
        }, done.fail)
      })

      it('should reject with a request error if retries <= 0', done => {
        const client = axios.create()

        setupResponses(client, [
          () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
        ])

        axiosRetry(client, { retries: 0, retryCondition: () => {} })

        client.get('http://example.com/test').then(done.fail, error => {
          expect(error).toBe(NETWORK_ERROR)
          done()
        })
      })

      it('should reject with a request error if there are more errors than retries', done => {
        const client = axios.create()

        setupResponses(client, [
          () => nock('http://example.com').get('/test').replyWithError(new Error('foo error')),
          () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
          () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
        ])

        axiosRetry(client, { retries: 2, retryCondition: () => true })

        client.get('http://example.com/test').then(done.fail, error => {
          expect(error).toBe(NETWORK_ERROR)
          done()
        })
      })

      it('should honor the original `timeout` across retries', done => {
        const client = axios.create()

        setupResponses(client, [
          () => nock('http://example.com').get('/test').delay(75).replyWithError(NETWORK_ERROR),
          () => nock('http://example.com').get('/test').delay(75).replyWithError(NETWORK_ERROR),
          () => nock('http://example.com').get('/test').reply(200),
        ])

        axiosRetry(client, { retries: 3 })

        client.get('http://example.com/test', { timeout: 100 }).then(() => {}, error => {
          expect(error.code).toBe('ECONNABORTED')
          done()
        }).catch(() => {})
      })

      it('should reject with errors without a `config` property without retrying', done => {
        const client = axios.create()

        setupResponses(client, [
          () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
          () => nock('http://example.com').get('/test').reply(200),
        ])

        // Force returning a plain error without extended information from Axios
        const generatedError = new Error()
        client.interceptors.response.use(null, () => Promise.reject(generatedError))

        axiosRetry(client, { retries: 1, retryCondition: () => true })

        client.get('http://example.com/test').then(done.fail, error => {
          expect(error).toBe(generatedError)
          done()
        })
      })

      it('should work with a custom `agent` configuration', done => {
        const httpAgent = new http.Agent()

        // Simulate circular structure
        const fakeSocket = { foo: 'foo' }
        httpAgent.sockets['multisearch.api.softonic.com:80:'] = [fakeSocket]
        fakeSocket.socket = fakeSocket

        const client = axios.create({ agent: httpAgent })
        setupResponses(client, [
          () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
          () => nock('http://example.com').get('/test').reply(200, 'It worked!'),
        ])

        axiosRetry(client, { retries: 1, retryCondition: () => true })

        client.get('http://example.com/test').then(result => {
          expect(result.status).toBe(200)
          done()
        }, done.fail)
      })

      it('should work with a custom `httpAgent` configuration', done => {
        const httpAgent = new http.Agent()

        // Simulate circular structure
        const fakeSocket = { foo: 'foo' }
        httpAgent.sockets['multisearch.api.softonic.com:80:'] = [fakeSocket]
        fakeSocket.socket = fakeSocket

        const client = axios.create({ httpAgent })
        setupResponses(client, [
          () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
          () => nock('http://example.com').get('/test').reply(200, 'It worked!'),
        ])

        axiosRetry(client, { retries: 1, retryCondition: () => true })

        client.get('http://example.com/test').then(result => {
          expect(result.status).toBe(200)
          done()
        }, done.fail)
      })
    })

    describe('when it does NOT satisfy the retry condition', () => {
      it('should reject with the error', (done) => {
        const client = axios.create()
        setupResponses(client, [
          () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
          () => nock('http://example.com').get('/test').reply(200, 'It worked!'),
        ])

        axiosRetry(client, { retries: 1, retryCondition: () => false })

        client.get('http://example.com/test').then(done.fail, (error) => {
          expect(error).toBe(NETWORK_ERROR)
          done()
        })
      })
    })
  })

  it('should use request-specific configuration', done => {
    const client = axios.create()

    setupResponses(client, [
      () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
      () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
      () => nock('http://example.com').get('/test').reply(200),
    ])

    axiosRetry(client, { retries: 0 })

    client.get('http://example.com/test', {
      'axios-retry': {
        retries: 2,
      },
    }).then(result => {
      expect(result.status).toBe(200)
      done()
    }, done.fail)
  })
})

describe('axiosRetry(axios, { retries, retryDelay })', () => {
  describe('when custom retryDelay function is supplied', () => {
    it('should execute for each retry', done => {
      const client = axios.create()

      setupResponses(client, [
        () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
        () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
        () => nock('http://example.com').get('/test').replyWithError(NETWORK_ERROR),
        () => nock('http://example.com').get('/test').reply(200, 'It worked!'),
      ])

      let retryCount = 0

      axiosRetry(client, {
        retries: 4,
        retryCondition: () => true,
        retryDelay: () => { retryCount += 1; return 0 },
      })

      client.get('http://example.com/test').then(() => {
        expect(retryCount).toBe(3)
        done()
      }, done.fail)
    })
  })
})

describe('isNetworkError(error)', () => {
  it('should be true for network errors like connection refused', () => {
    const connectionRefusedError = new Error()
    connectionRefusedError.code = 'ECONNREFUSED'
    expect(isNetworkError(connectionRefusedError)).toBe(true)
  })

  it('should be false for timeout errors', () => {
    const timeoutError = new Error()
    timeoutError.code = 'ECONNABORTED'
    expect(isNetworkError(timeoutError)).toBe(false)
  })

  it('should be false for errors with a response', () => {
    const responseError = new Error('Response error')
    responseError.response = { status: 500 }
    expect(isNetworkError(responseError)).toBe(false)
  })

  it('should be false for other errors', () => {
    expect(isNetworkError(new Error())).toBe(false)
  })
})

describe('isSafeRequestError(error)', () => {
  ['get', 'head', 'options'].forEach((method) => {
    it(`should be true for "${method}" requests with a 5xx response`, () => {
      const errorResponse = new Error('Error response')
      errorResponse.config = { method }
      errorResponse.response = { status: 500 }
      expect(isSafeRequestError(errorResponse)).toBe(true)
    })

    it(`should be true for "${method}" requests without a response`, () => {
      const errorResponse = new Error('Error response')
      errorResponse.config = { method }
      expect(isSafeRequestError(errorResponse)).toBe(true)
    })
  });

  ['post', 'put', 'patch', 'delete'].forEach((method) => {
    it(`should be false for "${method}" requests with a 5xx response`, () => {
      const errorResponse = new Error('Error response')
      errorResponse.config = { method }
      errorResponse.response = { status: 500 }
      expect(isSafeRequestError(errorResponse)).toBe(false)
    })

    it(`should be false for "${method}" requests without a response`, () => {
      const errorResponse = new Error('Error response')
      errorResponse.config = { method }
      expect(isSafeRequestError(errorResponse)).toBe(false)
    })
  })

  it('should be false for errors without a `config`', () => {
    const errorResponse = new Error('Error response')
    errorResponse.response = { status: 500 }
    expect(isSafeRequestError(errorResponse)).toBe(false)
  })

  it('should be false for non-5xx responses', () => {
    const errorResponse = new Error('Error response')
    errorResponse.config = { method: 'get' }
    errorResponse.response = { status: 404 }
    expect(isSafeRequestError(errorResponse)).toBe(false)
  })

  // eslint-disable-next-line jasmine/no-spec-dupes
  it('should be false for aborted requests', () => {
    const errorResponse = new Error('Error response')
    errorResponse.code = 'ECONNABORTED'
    errorResponse.config = { method: 'get' }
    expect(isSafeRequestError(errorResponse)).toBe(false)
  })
})

describe('isIdempotentRequestError(error)', () => {
  ['get', 'head', 'options', 'put', 'delete'].forEach((method) => {
    it(`should be true for "${method}" requests with a 5xx response`, () => {
      const errorResponse = new Error('Error response')
      errorResponse.config = { method }
      errorResponse.response = { status: 500 }
      expect(isIdempotentRequestError(errorResponse)).toBe(true)
    })

    it(`should be true for "${method}" requests without a response`, () => {
      const errorResponse = new Error('Error response')
      errorResponse.config = { method }
      expect(isIdempotentRequestError(errorResponse)).toBe(true)
    })
  });

  ['post', 'patch'].forEach((method) => {
    it(`should be false for "${method}" requests with a 5xx response`, () => {
      const errorResponse = new Error('Error response')
      errorResponse.config = { method }
      errorResponse.response = { status: 500 }
      expect(isIdempotentRequestError(errorResponse)).toBe(false)
    })

    it(`should be false for "${method}" requests without a response`, () => {
      const errorResponse = new Error('Error response')
      errorResponse.config = { method }
      errorResponse.response = { status: 500 }
      expect(isIdempotentRequestError(errorResponse)).toBe(false)
    })
  })

  // eslint-disable-next-line jasmine/no-spec-dupes
  it('should be false for errors without a `config`', () => {
    const errorResponse = new Error('Error response')
    errorResponse.response = { status: 500 }
    expect(isIdempotentRequestError(errorResponse)).toBe(false)
  })

  // eslint-disable-next-line jasmine/no-spec-dupes
  it('should be false for non-5xx responses', () => {
    const errorResponse = new Error('Error response')
    errorResponse.config = { method: 'get' }
    errorResponse.response = { status: 404 }
    expect(isIdempotentRequestError(errorResponse)).toBe(false)
  })

  // eslint-disable-next-line jasmine/no-spec-dupes
  it('should be false for aborted requests', () => {
    const errorResponse = new Error('Error response')
    errorResponse.code = 'ECONNABORTED'
    errorResponse.config = { method: 'get' }
    expect(isIdempotentRequestError(errorResponse)).toBe(false)
  })
})

describe('exponentialDelay', () => {
  it('should return exponential retry delay', () => {
    function assertTime(retryNumber) {
      const min = (Math.pow(2, retryNumber) * 100)
      const max = (Math.pow(2, retryNumber * 100) * 0.2)

      const time = exponentialDelay(retryNumber)

      expect((time >= min && time <= max)).toBe(true)
    }

    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].forEach(assertTime)
  })
})

describe('isRetryableError(error)', () => {
  // eslint-disable-next-line jasmine/no-spec-dupes
  it('should be false for aborted requests', () => {
    const errorResponse = new Error('Error response')
    errorResponse.code = 'ECONNABORTED'
    expect(isRetryableError(errorResponse)).toBe(false)
  })

  it('should be true for timeouts', () => {
    const errorResponse = new Error('Error response')
    errorResponse.code = 'ECONNRESET'
    expect(isRetryableError(errorResponse)).toBe(true)
  })

  it('should be true for a 5xx response', () => {
    const errorResponse = new Error('Error response')
    errorResponse.code = 'ECONNRESET'
    errorResponse.response = { status: 500 }
    expect(isRetryableError(errorResponse)).toBe(true)
  })

  it('should be false for a response !== 5xx', () => {
    const errorResponse = new Error('Error response')
    errorResponse.code = 'ECONNRESET'
    errorResponse.response = { status: 400 }
    expect(isRetryableError(errorResponse)).toBe(false)
  })
})

describe('axiosRetry(axios, {retryTimeout})', () => {
  describe('with concurrent requests', () => {
    it('should return the first request that responds ok', done => {
      const client = axios.create()

      setupResponses(client, [
        () => nock('http://example.com').get('/test').delay(150).reply(200, 'Response 1'),
        () => nock('http://example.com').get('/test').delay(150).reply(200, 'Response 2'),
      ])

      axiosRetry(client)

      client.get('http://example.com/test', { retryTimeout: 100 }).then(response => {
        expect(String(response.data)).toBe('Response 1')
        done()
      }, done.fail).catch(done.fail)
    })

    it('should return the first request that responds ok, even if it comes later', done => {
      const client = axios.create()

      setupResponses(client, [
        () => nock('http://example.com').get('/test').delay(3500).reply(200, 'Response 1'),
        () => nock('http://example.com').get('/test').delay(150).reply(200, 'Response 2'),
      ])

      axiosRetry(client)

      client.get('http://example.com/test', { retryTimeout: 100 }).then(response => {
        expect(String(response.data)).toBe('Response 2')
        done()
      }, done.fail).catch(done.fail)
    })

    it('should get an ok response after error', done => {
      const client = axios.create()

      const customError = {
        ...NETWORK_ERROR,
        message: 'falha',
      }

      setupResponses(client, [
        () => nock('http://example.com').get('/test').delay(200).replyWithError(customError),
        () => nock('http://example.com').get('/test').delay(200).reply(200, 'It worked!'),
      ])

      axiosRetry(client)

      client.get('http://example.com/test', { retryTimeout: 150 }).then(response => {
        expect(response.status).toBe(200)
        done()
      }, done.fail)
    })

    it('should get the first response after an error', done => {
      const client = axios.create()

      setupResponses(client, [
        () => nock('http://example.com').get('/test').delay(150).reply(200, 'It worked!'),
        () => nock('http://example.com').get('/test').delay(150).replyWithError(NETWORK_ERROR),
      ])

      axiosRetry(client)

      client.get('http://example.com/test', { retryTimeout: 100 }).then(response => {
        expect(response.status).toBe(200)
        expect(response.data).toBe('It worked!')
        done()
      }, done.fail)
    })
  })
})

describe('cancellation', () => {
  describe('with cancel token', () => {
    it('should abort', done => {
      const client = axios.create()

      setupResponses(client, [
        () => nock('http://example.com').get('/test').delay(150).reply(200, 'It worked!'),
      ])

      const cancelSource = axios.CancelToken.source()
      const cancelToken = cancelSource.token

      client.get('http://example.com/test', { cancelToken })
        .then(done.fail)
        .catch(error => {
          expect(error.constructor.name).toBe('Cancel')
          expect(error.message).toBe('user cancelled')
          done()
        })

      setTimeout(() => cancelSource.cancel('user cancelled'), 100)
    })
  })
})

describe('qwest legacy', () => {
  describe('timeout per request', () => {
    it('should use the timeout value per attempt instead of using it for all retries', done => {
      const client = axios.create()

      setupResponses(client, [
        () => nock('http://example.com').get('/test').delay(150).reply(200, 'First attempt'),
        () => nock('http://example.com').get('/test').delay(150).reply(200, 'Second attempt'),
        () => nock('http://example.com').get('/test').delay(50).reply(200, 'Third attempt'),
      ])

      axiosRetry(client, {
        retries: 5,
        qwestLegacy: true,
      })

      client.get('http://example.com/test', {
        timeout: 100,
      })
        .then(response => {
          expect(response.data).toBe('Third attempt')
          done()
        })
        .catch(done.fail)
    })
  })
})


index.d.ts
import * as axios from 'axios'

export interface IAxiosRetryConfig {
  /**
   * The number of times to retry before failing
   * default: 3
   * 
   * @type {number}
   */
  retries?: number,
  /**
   * A callback to further control if a request should be retried. By default, it retries if the result did not have a response.
   * default: error => !error.response
   * 
   * @type {Function}
   */
  retryCondition?: (error: axios.AxiosError) => boolean,
  /**
   * A callback to further control the delay between retry requests. By default there is no delay.
   *
   * @type {Function}
   */
  retryDelay?: (retryCount: number, error: axios.AxiosError) => number
}

export interface IAxiosRetry {
  (
    axios: axios.AxiosStatic | axios.AxiosInstance,
    axiosRetryConfig?: IAxiosRetryConfig
  );
  exponentialDelay(retryNumber?: number): number;
}

declare const axiosRetry: IAxiosRetry

export default axiosRetry


es/index.js
import isRetryAllowed from 'is-retry-allowed'
import axiosLib from 'axios'
import safeStringify from 'fast-safe-stringify'

const CancelToken = axiosLib.CancelToken

const namespace = 'axios-retry'

/**
 * @param  {Error}  error
 * @return {boolean}
 */
export function isNetworkError(error) {
  const {
    qwestLegacy = false,
  } = error.config || {}

  return !error.response &&
    Boolean(error.code) && // Prevents retrying cancelled requests
    (
      error.code !== 'ECONNABORTED' || !!qwestLegacy
    ) && // Prevents retrying timed out requests
    isRetryAllowed(error) // Prevents retrying unsafe errors
}

const SAFE_HTTP_METHODS = ['get', 'head', 'options']
const IDEMPOTENT_HTTP_METHODS = SAFE_HTTP_METHODS.concat(['put', 'delete'])

/**
 * @param  {Error}  error
 * @return {boolean}
 */
export function isRetryableError(error) {
  return error.code !== 'ECONNABORTED' &&
    (!error.response || (error.response.status >= 500 && error.response.status <= 599))
}

/**
 * @param  {Error}  error
 * @return {boolean}
 */
export function isSafeRequestError(error) {
  if (!error.config) {
    // Cannot determine if the request can be retried
    return false
  }

  return isRetryableError(error) && SAFE_HTTP_METHODS.indexOf(error.config.method) !== -1
}

/**
 * @param  {Error}  error
 * @return {boolean}
 */
export function isIdempotentRequestError(error) {
  if (!error.config) {
    // Cannot determine if the request can be retried
    return false
  }

  return isRetryableError(error) && IDEMPOTENT_HTTP_METHODS.indexOf(error.config.method) !== -1
}

/**
 * @param  {Error}  error
 * @return {boolean}
 */
export function isNetworkOrIdempotentRequestError(error) {
  return isNetworkError(error) || isIdempotentRequestError(error)
}

/**
 * @return {number} - delay in milliseconds, always 0
 */
function noDelay() {
  return 0
}

/**
 * @param  {number} [retryNumber=0]
 * @return {number} - delay in milliseconds
 */
export function exponentialDelay(retryNumber = 0) {
  const delay = Math.pow(2, retryNumber) * 100
  const randomSum = delay * 0.2 * Math.random() // 0-20% of the delay
  return delay + randomSum
}

/**
 * Initializes and returns the retry state for the given request/config
 * @param  {AxiosRequestConfig} config
 * @return {Object}
 */
function getCurrentState(config) {
  const currentState = config[namespace] || {}
  currentState.retryCount = currentState.retryCount || 0
  currentState.requests = currentState.requests || []
  currentState.cancelled = 'cancelled' in currentState ? currentState.cancelled : false
  currentState.response = 'response' in currentState ? currentState.response : null

  config[namespace] = currentState
  return currentState
}

/**
 * Returns the axios-retry options for the current request
 * @param  {AxiosRequestConfig} config
 * @param  {AxiosRetryConfig} defaultOptions
 * @return {AxiosRetryConfig}
 */
function getRequestOptions(config, defaultOptions) {
  return Object.assign({}, defaultOptions, config[namespace])
}

/**
 * @param  {Axios} axios
 * @param  {AxiosRequestConfig} config
 */
function fixConfig(axios, config) {
  if (axios.defaults.agent === config.agent) {
    delete config.agent
  }
  if (axios.defaults.httpAgent === config.httpAgent) {
    delete config.httpAgent
  }
  if (axios.defaults.httpsAgent === config.httpsAgent) {
    delete config.httpsAgent
  }
}

function createCancelToken() {
  const cancelSource = CancelToken.source()
  return {
    cancelSource,
    cancelToken: cancelSource.token,
  }
}

function removeCircular(obj) {
  if (obj) {
    return JSON.parse(safeStringify(obj))
  }
  return obj
}

function cleanupOutput(output) {
  const keys = Object.keys(output)
  keys.forEach(key => {
    output[key] = removeCircular(output[key])
  })
  return output
}

/**
 * Adds response interceptors to an axios instance to retry requests failed due to network issues
 *
 * @example
 *
 * import axios from 'axios';
 *
 * axiosRetry(axios, { retries: 3 });
 *
 * axios.get('http://example.com/test') // The first request fails and the second returns 'ok'
 *   .then(result => {
 *     result.data; // 'ok'
 *   });
 *
 * // Exponential back-off retry delay between requests
 * axiosRetry(axios, { retryDelay : axiosRetry.exponentialDelay});
 *
 * // Custom retry delay
 * axiosRetry(axios, { retryDelay : (retryCount) => {
 *   return retryCount * 1000;
 * }});
 *
 * // Also works with custom axios instances
 * const client = axios.create({ baseURL: 'http://example.com' });
 * axiosRetry(client, { retries: 3 });
 *
 * client.get('/test') // The first request fails and the second returns 'ok'
 *   .then(result => {
 *     result.data; // 'ok'
 *   });
 *
 * // Allows request-specific configuration
 * client
 *   .get('/test', {
 *     'axios-retry': {
 *       retries: 0
 *     }
 *   })
 *   .catch(error => { // The first request fails
 *     error !== undefined
 *   });
 *
 * @param {Axios} axios An axios instance (the axios object or one created from axios.create)
 * @param {Object} [defaultOptions]
 * @param {number} [defaultOptions.retries=3] Number of retries
 * @param {Function} [defaultOptions.retryCondition=isNetworkOrIdempotentRequestError]
 *        A function to determine if the error can be retried
 * @param {Function} [defaultOptions.retryDelay=noDelay]
 *        A function to determine the delay between retry requests
 * @return {undefined}
 */
export default function axiosRetry(axios, defaultOptions) {
  const cancelAll = (config) => {
    const currentState = getCurrentState(config)

    currentState.cancelled = true

    const { requests } = currentState

    requests
      .forEach(request => {
        request.preventRetryFromTimeout = true
        request.preventRetryFromError = true
      })

    if (config.cancelSource) {
      config.cancelSource.cancel({ config })
    }
  }

  const retry = (config) => {
    const request = axios({
      ...config,
    })
    request.catch(() => {})

    return request
  }

  axios.interceptors.request.use((config) => {
    const currentState = getCurrentState(config)
    currentState.lastRequestTime = Date.now()

    const { retryTimeout, cancelToken } = config

    let curConfig = { ...config }

    if (!cancelToken) {
      curConfig = {
        ...curConfig,
        ...createCancelToken(),
      }
    }

    const {
      qwestLegacy = false,
    } = getRequestOptions(config, defaultOptions)

    curConfig = {
      ...curConfig,
      qwestLegacy,
      preventRetryFromTimeout: false,
      preventRetryFromError: false,
      hasRetriedFromTimeout: false,
      hasRetriedFromError: false,
      retryFromTimeout: null,
    }

    if (retryTimeout) {
      setTimeout(() => {
        if (!curConfig.preventRetryFromTimeout) {
          curConfig.preventRetryFromError = true
          curConfig.hasRetriedFromTimeout = true

          curConfig.retryFromTimeout = retry({ ...curConfig })
        }
      }, retryTimeout)
    }

    currentState.requests.push(curConfig)

    return curConfig
  })

  // In case of success
  axios.interceptors.response.use(response => {
    const returnValue = Promise.resolve(response)

    const { config } = response
    if (!config) return returnValue

    const currentState = getCurrentState(config)
    if (!currentState) return returnValue

    currentState.response = response

    // cancel all pending requests
    cancelAll(config)

    return returnValue
  }, error => {
    // In case of error

    const isCancellation = error.__CANCEL__
    const config = isCancellation && error.message && error.message.config ? error.message.config : error.config

    // If we have no information to retry the request
    if (!config) {
      return Promise.reject(cleanupOutput(error))
    }

    const currentState = getCurrentState(config)

    if (currentState && currentState.response) {
      return Promise.resolve(currentState.response)
    }

    const {
      retries = 3,
      retryCondition = isNetworkOrIdempotentRequestError,
      retryDelay = noDelay,
      qwestLegacy = false,
    } = getRequestOptions(config, defaultOptions)

    const retryMeetsCondition = retryCondition(error)
    const shouldRetry =
      retryMeetsCondition &&
      currentState.retryCount < retries &&
      (!config.hasTimedOut || qwestLegacy) &&
      !config.preventRetryFromError

    config.preventRetryFromTimeout = true

    if (shouldRetry) {
      currentState.retryCount++
      const delay = retryDelay(currentState.retryCount, error)

      // Axios fails merging this configuration to the default configuration because it has an issue
      // with circular structures: https://github.com/mzabriskie/axios/issues/370
      fixConfig(axios, config)

      if (!qwestLegacy && config.timeout && currentState.lastRequestTime) {
        const lastRequestDuration = Date.now() - currentState.lastRequestTime
        // Minimum 1ms timeout (passing 0 or less to XHR means no timeout)
        config.timeout = Math.max((config.timeout - lastRequestDuration) - delay, 1)
        if (config.timeout === 1) {
          config.hasTimedOut = true
        }
      }

      return new Promise((resolve, reject) => {
        setTimeout(() => {
          // retries if it should continue with retrying from error
          if (!config.preventRetryFromError) {
            config.hasRetriedFromError = true
            config.preventRetryFromTimeout = true
            return retry(config)
              .then(result => {
                resolve(result)
              })
              .catch(e => {
                reject(cleanupOutput(e))
              })
          } else if (config.hasRetriedFromTimeout) {
            // if in this meantime (i.e. the delay from retrying from error)
            // the retry from timeout has been dispatched, return that
            // instead and avoid retrying
            return config.retryFromTimeout
              .then(result => resolve(result))
              .catch(e => {
                reject(cleanupOutput(e))
              })
          }
        }, delay)
      })
    // if it still can retry, but retry from error has been prevented,
    // and it has retried from timeout
    } else if (retryMeetsCondition && !config.hasTimedOut && config.hasRetriedFromTimeout) {
      return config.retryFromTimeout
    }

    // reject if can't retry anymore
    return Promise.reject(cleanupOutput(error))
  })
}

// Compatibility with CommonJS
axiosRetry.isNetworkError = isNetworkError
axiosRetry.isSafeRequestError = isSafeRequestError
axiosRetry.isIdempotentRequestError = isIdempotentRequestError
axiosRetry.isNetworkOrIdempotentRequestError = isNetworkOrIdempotentRequestError
axiosRetry.exponentialDelay = exponentialDelay
